API for providing untrusted data about unverified signatures

mtrmac opened the pull request:
@mfojtik PTAL

This utlimately adds `signature.UntrustedSignatureInformation` and `signature.GetUntrustedSignatureInformationWithoutVerifying`; https://github.com/mtrmac/skopeo/blob/unverified-contents/cmd/skopeo/signing.go#L93 is an example user, with output like
```json
{
    "UntrustedDockerManifestDigest": "sha256:a59906e33509d14c036c8678d687bd4eec81ed7c4b8ce907b888c607f6a1e0e6",
    "UntrustedDockerReference": "mtrmac/busybox-test:latest",
    "UntrustedCreatorID": "atomic 0.1.0-dev",
    "UntrustedTimestamp": 1472502383,
    "UntrustedShortKeyIdentifier": "5A33F660B38479DF"
}
```

(For a real code review, see the individual commit messages for detailed descriptions. But right now this is missing tests, so please do not merge.)

mtrmac commented on signature/signature.go:
```diff
@@ -27,37 +27,70 @@ func (err InvalidSignatureError) Error() string {
+	UntrustedDockerManifestDigest digest.Digest
+	UntrustedDockerReference      string // FIXME: more precise type?
+	UntrustedCreatorID            *string
+	UntrustedTimestamp            *int64
```
Should this be a stupid `int64` or a `time.Time`? `time.Time` would be nicer to use at the cost of a _tiny_ risk that the `int64`→`Time` conversion might be vulnerable (though looking at the implementation that seems _extremely_ unlikely, for the _current_ implementation at least).

mtrmac replied:
Actually, there is no risk/tradeoff here; we can do the `UnmarshalJSON` into an `int64` in the private `untrustedSignature`, and do the conversion to `time.Time` in `GetUntrustedSignatureInformationWithoutVerifying`. Changed like that.

mfojtik replied:
I still think this should be `*time.Time` so the consumers life will be easier (they don't have to convert this to a time which is what you want to display to users).

mtrmac replied:
Right now the only consumer is `GetUntrustedSignatureInformationWithoutVerifying`, so placement of the conversion makes no difference to the total complexity of the package.

Using the native `int64` here is nice in that it explicitly corresponds to the JSON semantics, and we are more certain that we can round-trip the value and that the caller setting the value is aware about the conversion (e.g., compared to `time.Time`, that this records neither timezones nor sub-second resolutions).

Though, perhaps, before we start using this field in earnest and before we ship the first version of a validating parser of this fields, perhaps we should change the type to a `float64` (with its sub-second resolution and a more restricted precision), which is what JSON _actually_ stores? The `int64` also has conversion issues. @runcom  ?

mfojtik commented:
I will be better to have time.Time as it will need to be converted anyway
to store as api field.

-- 
Michal Fojtik

On 11 January 2017 at 20:55:42, Miloslav Trmač (notifications@github.com)
wrote:

> *@mtrmac* commented on this pull request.
> ------------------------------
>
> In signature/signature.go
> <https://github.com/containers/image/pull/208#pullrequestreview-16221329>:
>
> >  type Signature struct {
>         DockerManifestDigest digest.Digest
>         DockerReference      string // FIXME: more precise type?
>  }
>
> -// Wrap signature to add to it some methods which we don't want to make public.
> -type privateSignature struct {
> -       Signature
> +// untrustedSignature is a parsed content of a signature.
> +type untrustedSignature struct {
> +       UntrustedDockerManifestDigest digest.Digest
> +       UntrustedDockerReference      string // FIXME: more precise type?
> +       UntrustedCreatorID            *string
> +       UntrustedTimestamp            *int64
>
> Should this be a stupid int64 or a time.Time? time.Time would be nicer to
> use at the cost of a *tiny* risk that the int64→Time conversion might be
> vulnerable (though looking at the implementation that seems *extremely*
> unlikely, for the *current* implementation at least).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/containers/image/pull/208#pullrequestreview-16221329>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACsaFJrN6YOUEIgOOKuwGtiYS5bo4oeks5rRTO-gaJpZM4Lg_xY>
> .
>

mfojtik commented on signature/signature.go:
```diff
@@ -27,37 +27,70 @@ func (err InvalidSignatureError) Error() string {
+// There is NO REASON to expect the values to be correct, or not intentionally misleading
+// (including things like “✅ Verified by $authority”)
+type UntrustedSignatureInformation struct {
+	UntrustedDockerManifestDigest digest.Digest
```
Strip the `Untrusted` prefix from the fields (the struct is enough)

mtrmac replied:
The type name is invisible with `info, err := …`; and the point of use of the structure and the point of calling the producer of the value may be pretty far apart.

mfojtik replied:
I think it is more Go idiomatic to not repeat the prefix if the struct in the functions that struct define. The callers should use this like:

```
untrustedInfo, err := ....
untrustedInfo.DockerManifestDigest()
```

WDYT?

mtrmac replied:
Honestly I am fairly uncomfortable with providing the unverified information because I think there is a pretty big risk either code, or, more likely, humans will trust it anyway (while I do realize that providing it is unavoidable), and naming the fields this way is my last line of defense against inadvertent mistakes. Of course nothing forces any consumer to use these names in an UI.

mfojtik commented on signature/signature.go:
```diff
@@ -27,37 +27,70 @@ func (err InvalidSignatureError) Error() string {
+	UntrustedDockerManifestDigest digest.Digest
+	UntrustedDockerReference      string // FIXME: more precise type?
+	UntrustedCreatorID            *string
+	UntrustedTimestamp            *int64
```
This should be really *time.Time

mfojtik commented on signature/json.go:
```diff
@@ -29,6 +29,23 @@ func validateExactMapKeys(m map[string]interface{}, expectedKeys ...string) erro
+func int64Field(m map[string]interface{}, fieldName string) (int64, error) {
+	untyped, ok := m[fieldName]
+	if !ok {
+		return -1, jsonFormatError(fmt.Sprintf("Field %s missing", fieldName))
```
no need to return `-1` here. the 0 is default for int64 when not set.

mtrmac replied:
Well it is returning -1, not 0. Anyway, there is no “good” undefined value to return here if the caller makes a mistake and ignores `err`. -1 is perhaps _slightly_ less likely to be valid than a non-negative number… but it really is anybody’s guess whether that makes it more likely to be noticed as invalid or more likely to have completely unpredicted effects. (Go’s error handling sucks, it would be nice to have something like optional checked exceptions, so that a caller can’t ignore an error by mistake without it being very visible.)

It seems to me not to be worth it to bother with named return parameters so that we can skip explicitly naming a value returned in the error case—especially because that _also_ makes a decision, namely to use 0, only that decision is not explicitly visible.

mfojtik replied:
@mtrmac the called should never ignore "err", doing so is a programmer error and we can't prevent all human errors :-)

I'm fine with `-1`, just pointing out it is something unusual to me as usually if the return value is not defined, we default to a default value of the type (which in case of int64 is '0').

mtrmac replied:
> @mtrmac the called should never ignore "err", doing so is a programmer error and we can't prevent all human errors :-)

True, and yet, we are in the business of engineering tools and interfaces to minimize errors anyway. (Ordinarily I wouldn’t bother, but the `signature` subpackage brings out my worst pedantic tendencies.)

mfojtik commented:
LGTM

@mtrmac @runcom is there something left in this PR (that cannot be a follow up?)

mtrmac commented:
Updated:

- Added tests
- Separated unrelated trivial fixes into #213.
- (Tried making the timestamp a `float64`, but that doesn't have enough precision to represent nanosecond timestamps, so this turned out to be a dead end.)

@mfojtik @runcom PTAL, this is now ready for review.

mfojtik commented on signature/mechanism.go:
```diff
@@ -4,9 +4,13 @@ package signature
+	"strings"
 
 	"github.com/mtrmac/gpgme"
+	"golang.org/x/crypto/openpgp"
```
this wont work for openshift, we will not build with gpgme :(

mtrmac replied:
This does not add any _new_ uses of `gpgme`. Getting rid of gpgme happens in #207 not here (though the two do touch the same code, so the latter one will have to be modified. I expect the newly added `UntrustedSignatureContents` to remain in `mechanism.go` as a shared helper, called by both GPG backends.)

runcom commented:
:+1:

mtrmac commented:
Tests pass in https://github.com/projectatomic/skopeo/pull/293.

runcom commented:
Feel free to merge

mtrmac commented:
:+1:
